docs/design/trimmed-assemblies.md

```diff
@@ -0,0 +1,183 @@
+
+### `ManagedAssemblyToLink`
+The `PrepareForILLink` has a dependency that creates the ItemGroup `ManagedAssemblyToLink`, which represents the set of assemblies that will be passed to the linker. Custom targets may modify `IsTrimmable` and `TrimMode` metadata on these assemblies before `PrepareForILLink`, which will set the assembly action based on this metadata, or they may modify the metadata after `PrepareForILLink` has run.
+
+It will be illegal to change the items in `ManagedAssemblyToLink`, since this represents the set that needs to be filtered and replaced in the publish output. To change which assemblies are passed to the linker, a different extension point should be used to set `PostProcessAssemblies` metadata.
+
+### Examples
+
+This shows how a developer could turn on aggressive trimming for framework assemblies (which are defined to be `IsTrimmable` by the SDK):
+
+```xml
+<PropertyGroup>
+  <TrimMode>link</TrimMode>
+</PropertyGroup>
+```
+
+This shows how Blazor (or a developer) could hook into the build to opt assemblies into different levels of trimming based on the filename:
+
+```xml
+<Target Name="PrepareForBlazorILLink"
+        BeforeTargets="PrepareForILLink">
+  <PropertyGroup>
+    <!-- Set the default TrimMode for IsTrimmable assemblies -->
+    <TrimMode>link</TrimMode>
+  </PropertyGroup>
+  <ItemGroup>
+    <ManagedAssemblyToLink Condition="'$([System.String]::Copy('%(ManagedAssemblyToLink.Filename)').StartsWith('Microsoft.AspNetCore.'))">
+      <!-- Trim these assemblies using the global TrimMode -->
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+    <ManagedAssemblyToLink Condition="'$([System.String]::Copy('%(ManagedAssemblyToLink.Filename)').StartsWith('ThirdPartyAssembly.'))">
+      <!-- Trim these assemblies with assembly-level trimming. Implies IsTrimmable. -->
+      <TrimMode>copyused</TrimMode>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+### Other options
+
+.NET 5 introduced a host of additional SDK options that map directly to the underlying illink options. The full list is documented at https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming-options.
+
+## .NET 6
+
+### `IsTrimmable` `AssemblyMetadataAttribute`
+
+An assembly-level `AssemblyMetadataAttribute` may be placed on an assembly to indicate that it should be trimmed:
+
+```csharp
+[AssemblyMetadataAttribute("IsTrimmable", "true")]
+```
+
+The behavior is the same as the `IsTrimmable` MSBuild metadata, so that:
+- Assemblies with this attribute are trimmed with the global `TrimMode`
+- Assemblies without this attribute are rooted, and given the `copy` action
+
+The only understood value is `true`. Adding `[AssemblyMetadataAttribute("IsTrimmable", "false")]` will have no effect on the linker's behavior, because unattributed assemblies are assumed not to be trimmable by default. We will issue a warning in this case, to discourage misleading use of the attribute.
```

Comment:
> We will issue a warning in this case, to discourage misleading use of the attribute.

Should we allow that to make that symmetric with MSBuild metadata behaviour?

Reply:
This might make more sense if the defaults change - I mentioned it in the Future section.